src/main/java/net/snowflake/ingest/streaming/internal/ExternalVolumeManager.java

```diff
@@ -0,0 +1,172 @@
+  // Reference to the external volume per table
+  private final Map<String, StreamingIngestStorage<T, ExternalVolumeLocation>> externalVolumeMap;
+
+  // name of the owning client
+  private final String clientName;
+
+  // role of the owning client
+  private final String role;
+
+  // Reference to the Snowflake service client used for configure calls
+  private final SnowflakeServiceClient snowflakeServiceClient;
+
+  // Client prefix generated by the Snowflake server
+  private final String clientPrefix;
+
+  /**
+   * Constructor for ExternalVolumeManager
+   *
+   * @param isTestMode whether the manager in test mode
+   * @param role the role of the client
+   * @param clientName the name of the client
+   * @param snowflakeServiceClient the Snowflake service client used for configure calls
+   */
+  ExternalVolumeManager(
+      boolean isTestMode,
+      String role,
+      String clientName,
+      SnowflakeServiceClient snowflakeServiceClient) {
+    this.role = role;
+    this.clientName = clientName;
+    this.snowflakeServiceClient = snowflakeServiceClient;
+    this.externalVolumeMap = new ConcurrentHashMap<>();
+    try {
+      this.clientPrefix =
+          isTestMode
+              ? "testPrefix"
+              : this.snowflakeServiceClient
+                  .clientConfigure(new ClientConfigureRequest(role))
+                  .getClientPrefix();
+    } catch (IngestResponseException | IOException e) {
+      throw new SFException(e, ErrorCode.CLIENT_CONFIGURE_FAILURE, e.getMessage());
+    }
+  }
+
+  /**
+   * Given a channel context, return the target storage by looking up the table name
+   *
+   * @param channelFlushContext the channel flush context containing the table name
+   * @return target storage
+   */
+  @Override
+  public StreamingIngestStorage<T, ExternalVolumeLocation> getStorage(
+      ChannelFlushContext channelFlushContext) {
+    // Only one chunk per blob in Iceberg mode.
+    StreamingIngestStorage<T, ExternalVolumeLocation> stage =
+        this.externalVolumeMap.get(channelFlushContext.getFullyQualifiedTableName());
+
+    if (stage == null) {
+      throw new SFException(
+          ErrorCode.INTERNAL_ERROR,
+          String.format(
+              "No storage found for table %s", channelFlushContext.getFullyQualifiedTableName()));
+    }
+
+    return stage;
+  }
+
+  /**
+   * Add a storage to the manager by looking up the table name from the open channel response
+   *
+   * @param dbName the database name
+   * @param schemaName the schema name
+   * @param tableName the table name
+   * @param fileLocationInfo response from open channel
+   */
+  @Override
+  public void addStorage(
+      String dbName, String schemaName, String tableName, FileLocationInfo fileLocationInfo) {
+    String fullyQualifiedTableName = String.format("%s.%s.%s", dbName, schemaName, tableName);
+
+    try {
+      this.externalVolumeMap.putIfAbsent(
```

Comment:
Should we use computeIfAbsent? Also this means that we're that going to update the external volume if there is an entry? Is that correct? Think about a case where customer drops and recreates the same table name with a different external volume, then this would continue upload to the old volume?

Reply:
Ah yes this is a problem. If a customer drops and recreates a table with the same name but a different external volume, should all channel flushing to that table be invalidated or continue to flush to the new storage? If we choose to flush to the new storage, should we clear all the data that hasn't been flushed yet?

cc: @sfc-gh-hmadan

Reply:
good catch! lets track this with a JIRA? I'm leaning towards invalidating old data as a customer doing a drop and recreate is doing a destructive operation in the first place.

Reply:
[Jira](https://snowflakecomputing.atlassian.net/browse/SNOW-1527938) created. Changed to `put` now as the SDK currently continue streaming to the new table without invalidate old channels or clear channels buffer.